lucene/core/src/java/org/apache/lucene/search/QueryVisitor.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.search;
+
+import java.util.Set;
+
+import org.apache.lucene.index.Term;
+
+/**
+ * Allows recursion through a query tree
+ *
+ * @see Query#visit(QueryVisitor)
+ */
+public abstract class QueryVisitor {
+
+  /**
+   * Called by leaf queries that match on a specific term
+   *
+   * @param query the leaf query
+   * @param term  the term the query will match on
+   */
+  public void consumesTerm(Query query, Term term) { }
```

Comment:
consume or consumes ?

Reply:
is the query argument needed? I also wonder, do terms need to be provided one by one, or could the term argument be a varargs ? just an idea, not sure it makes sense

Reply:
I though about this too, I think a varArg API would be great but I think we also need a Iterator based API then? maybe not. Still varArg would be great we can implement early termination there easily on per-visitor and don't need to iterate everywhere.

Reply:
also +1 for singular vs. plural

Reply:
The tricky part with varArgs is what the semantics are - is it always the equivalent of a SHOULD or can PhraseQuery use it too, in which case it's a MUST?

Reply:
doesn't this depend on the SubVisitor here?